main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1694 @@
+object FastBoolUnification {
+
+  /**
+    * The threshold for when a solution is considered too complex.
+    *
+    * The threshold is specified as an upper bound on the permitted connectives in a substitution.
+    *
+    * The threshold must be at least 600 for the test suite to pass.
+    *
+    * If a solution is too complex an [[TooComplexException]] exception is thrown.
+    */
+  private val Threshold: Int = 1_000
+
+  /**
+    * Enable debugging (prints information during Boolean unification).
+    */
+  private val Debugging: Boolean = true
+
+  /**
+    * Internal formatter. Used for debugging.
+    */
+  private val formatter: Formatter = Formatter.NoFormatter
+
+  /**
+    * Attempts to solve all the given unification equations `l`.
+    *
+    * Returns `Ok(s)` where `s` is a most-general unifier for all equations.
+    *
+    * Returns `Err(c, l, s)` where `c` is a conflict, `l` is a list of unsolved equations, and `s` is a partial substitution.
+    */
+  def solveAll(l: List[Equation]): Result[BoolSubstitution, (FastBoolUnificationException, List[Equation], BoolSubstitution)] = {
+    val solver = new Solver(l)
+    solver.solve()
+  }
+
+  /**
+    * A stateful phased solver for Boolean unification equations. The solver maintains two fields that change over time:
+    *
+    * - The current (pending) unification equations to solve.
+    * - The current (partial) substitution which represents the solution.
+    *
+    * @param l The input list of Boolean unification equations to solve.
+    */
+  private class Solver(l: List[Equation]) {
+
+    /**
+      * The current (pending) equations to solve.
+      *
+      * The list of pending equations decrease as the solver progresses.
+      *
+      * Note that the pending equations is not a strict subset of the original equations because they may be simplified during computation.
+      *
+      * If Boolean unification is successful, the list of pending equations will become empty at the end of the computation.
+      */
+    private var currentEqns: List[Equation] = l
+
+    /**
+      * The current substitution. Initially empty, but grows during computation.
+      */
+    private var currentSubst: BoolSubstitution = BoolSubstitution.empty
+
+    /**
+      * Attempts to solve the equation system that this solver was instantiated with.
+      *
+      * Returns `Result.Ok(s)` with a most-general substitution which solves the equations system.
+      *
+      * Returns `Result.Err((ex, l, s))` where `c` is a conflict, `l` is a list of unsolved equations, and `s` is a partial substitution.
+      */
+    def solve(): Result[BoolSubstitution, (FastBoolUnificationException, List[Equation], BoolSubstitution)] = {
+      try {
+        phase0Init()
+        phase1UnitPropagation()
+        phase2VarPropagation()
+        phase3VarAssignment()
+        phase4SVE()
+        // verifySolution() // Not enabled by default due to cost. Enable for debugging.
+        verifySolutionSize()
```

Comment:
should also be commented out / conditionally run?

Reply:
This is what triggers TooComplex exception.

Reply:
So it should be there.